File: ds3-autogen-net/src/main/java/com/spectralogic/ds3autogen/net/generators/parsers/type/JobListParserGenerator.java
/*
 * ******************************************************************************
 *   Copyright 2014-2015 Spectra Logic Corporation. All Rights Reserved.
 *   Licensed under the Apache License, Version 2.0 (the "License"). You may not use
 *   this file except in compliance with the License. A copy of the License is located at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file.
 *   This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 *   CONDITIONS OF ANY KIND, either express or implied. See the License for the
 *   specific language governing permissions and limitations under the License.
 * ****************************************************************************
 */

package com.spectralogic.ds3autogen.net.generators.parsers.type;

import com.google.common.collect.ImmutableList;
import com.spectralogic.ds3autogen.api.models.apispec.Ds3Element;
import com.spectralogic.ds3autogen.net.generators.parsers.element.NullableElement;

import static com.spectralogic.ds3autogen.net.utils.NetNullableElementUtils.createNullableElement;
import static com.spectralogic.ds3autogen.utils.ConverterUtil.isEmpty;
import static com.spectralogic.ds3autogen.utils.Ds3ElementUtil.getXmlTagName;
import static com.spectralogic.ds3autogen.utils.Ds3ElementUtil.isAttribute;

/**
 * Generates the Type Parser model for the JobList type (i.e. JobsApiBean)
 */
public class JobListParserGenerator extends BaseTypeParserGenerator {

    /**
     * Converts a list of Ds3Elements into a list of Nullable Element
     */
    @Override
    public ImmutableList<NullableElement> toNullableElementsList(
            final ImmutableList<Ds3Element> ds3Elements,
            final boolean isObjectsType) {
        if (isEmpty(ds3Elements)) {
            return ImmutableList.of();
        }
        final ImmutableList.Builder<NullableElement> builder = ImmutableList.builder();
        for (final Ds3Element ds3Element : ds3Elements) {
            if (ds3Element.getName().equals("Jobs")) {
                builder.add(toJobsNullableElement(ds3Element));
            } else {
                builder.add(toNullableElement(ds3Element, isObjectsType));
            }
        }
        return builder.build();
    }

    /**
     * Converts the Jobs Ds3Element into the Jobs Nullable Element.
     * Assumption: ds3Element describes the Jobs element
     */
    protected static NullableElement toJobsNullableElement(final Ds3Element ds3Element) {
        if (!ds3Element.getName().equals("Jobs")) {
            throw new IllegalArgumentException("Cannot generate the Jobs Nullable element from the Ds3Element: " + ds3Element.getName());
        }
        return createNullableElement(
                toNullableElementName(ds3Element.getName(), false),
                ds3Element.getType(),
                ds3Element.getComponentType(),
                ds3Element.getNullable(),
                getXmlTagName(ds3Element),
                null,
                isAttribute(ds3Element.getDs3Annotations()));
    }
}

File: ds3-autogen-go/src/main/kotlin/com/spectralogic/ds3autogen/go/generators/parser/JobListParserGenerator.kt
/*
 * ******************************************************************************
 *   Copyright 2014-2017 Spectra Logic Corporation. All Rights Reserved.
 *   Licensed under the Apache License, Version 2.0 (the "License"). You may not use
 *   this file except in compliance with the License. A copy of the License is located at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file.
 *   This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 *   CONDITIONS OF ANY KIND, either express or implied. See the License for the
 *   specific language governing permissions and limitations under the License.
 * ****************************************************************************
 */

package com.spectralogic.ds3autogen.go.generators.parser

import com.google.common.collect.ImmutableMap
import com.spectralogic.ds3autogen.api.models.apispec.Ds3Element
import com.spectralogic.ds3autogen.api.models.apispec.Ds3Type
import com.spectralogic.ds3autogen.go.models.parser.ParseChildNodeAddToSlice
import com.spectralogic.ds3autogen.go.models.parser.ParseElement
import com.spectralogic.ds3autogen.go.utils.toGoType
import com.spectralogic.ds3autogen.api.models.apispec.decapitalize
import com.spectralogic.ds3autogen.api.models.apispec.capitalize

/**
 * The Go generator for JobList parser. This is special-cased because there is
 * no encapsulating data tag for the response payload represented by the type.
 * As a result, the annotations for an encapsulating tag produce incorrect code.
 */
class JobListParserGenerator : BaseTypeParserGenerator() {

    /**
     * Converts a Ds3Element into a ParsingElements. The Ds3Element with name "Job"
     * is special-cased to ignore the encapsulating tag "Job".
     */
    override fun toChildNode(ds3Element: Ds3Element, typeName: String, typeMap: ImmutableMap<String, Ds3Type>): ParseElement {
        if (ds3Element.name == "Jobs") {
            val xmlTag = getXmlTagName(ds3Element)
            val modelName = typeName.decapitalize()
            val paramName = ds3Element.name.capitalize()
            val childType = toGoType(ds3Element.componentType!!)
            return ParseChildNodeAddToSlice(xmlTag, modelName, paramName, childType)
        }

        return toStandardChildNode(ds3Element, typeName, typeMap)
    }
}